### PR TITLE
Improve sampling time by using a shared S3 client

### PR DESCRIPTION
**Pull request by derekluan**
**Goal:** Reduce runtime of sampling from all partitions.

**Solution:** Previously, code was creating a new s3 client for each get request, leading to a lot of overhead. Using a shared s3 client for get requests from the same task resulted in a big speed up.

**Rationale:** This is a safe approach because the boto s3 client is thread-safe: https://boto3.amazonaws.com/v1/documentation/api/latest/guide/clients.html.

**Final Results**
Original Raysort (with default maxworkers) with skewed data

- Sample All: 22.484s
- https://wandb.ai/raysort/raysort/runs/ryo9bymu

Shared s3 client Raysort (with default maxworkers + shared s3) with skewed data:

- Sample All: 2.531s
- https://wandb.ai/raysort/raysort/runs/1svdogy9

### CODE REVIEW

File: raysort/config/aws.py [hunk 1]
```diff
@@ -412,6 +412,25 @@
...
+            instance_count=10,
+            instance_type=i4i_2xl,
+        ),
+        system=dict(),
+        app=dict(
+            **get_steps(),
+            total_gb=1000,
+            input_part_gb=2,
+            s3_buckets=get_s3_buckets(),
+            map_parallelism_multiplier=1,
```
Comment (franklsf95) [inline raysort/config/aws.py:428]: Try reducing this to 0.75

File: scripts/misc/s3_generate_input.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
+"""
```
Comment (franklsf95) [inline scripts/misc/s3_generate_input.py:1]: We probably don't need this script since this is just `STEPS=generate_input python main.py`.

**Review by franklsf95 — APPROVED** [review]
Great work identifying the issue and fixing it!

File: raysort/config/common.py [hunk 3]
```diff
@@ -169,6 +169,10 @@ class AppConfig:
     data_dirs: list[str] = field(default_factory=list)
     is_local_cluster: bool = False
 
+    # runtime testing
+    # TODO (dl): do we need this to be configurable?
+    num_samples_per_partition: int = 20
```
[the remark this answers is not in the document]
Comment (derekluan) [reply raysort/config/common.py:174]: i just tested and it's already configurable by specifying in aws.py so it should be good

File: raysort/s3_utils.py [hunk 4]
```diff
@@ -97,10 +97,13 @@ def download(
     return np.frombuffer(buf.getbuffer(), dtype=np.uint8)
 
 
-def get_object_range(pinfo: PartInfo, bytes_range: tuple[int, int]) -> bytes:
+# s3_client is thread safe but not process safe, and should be reused if possible to reduce runtime.
```
[the remark this answers is not in the document]
Comment (derekluan) [reply raysort/s3_utils.py:100]: updated description

**Pull request merged**